File: q51-75/q55.py
# 2020.08.23
# Problem Statement:
# https://leetcode.com/problems/jump-game/

class Solution:
    def canJump(self, nums: List[int]) -> bool:
        # almost the same as question 45.
        
        # check corner cases
        if len(nums) == 1: return True
        
        steps = 1
        max_coverage = nums[0]
        pre_max = 0
        
        while True:
            # if already reach the end, return True
            if max_coverage >= len(nums) - 1:
                return True
            else:
                temp = max(i + nums[i] for i in range(pre_max, max_coverage+1))
                pre_max, max_coverage = max_coverage, temp
                # if after update, the max coverage is the same, cannot go further any more, return False
                if pre_max == max_coverage: return False
                steps = steps + 1

File: Amazon/q55.java
// 2022.05.27
// Problem Statement:
// https://leetcode.com/problems/jump-game/

// idea: expanding the max idx it can reach, 
// if the current index is already out of reach, directly return false,
// otherwise check if the max region includes the last position
class Solution {
    public boolean canJump(int[] nums) {
        int max = 0;
        for (int i=0; i<nums.length; i++) { // no need to exclude the last
            // when i==nums.length-1 (reaches the last), 
            // just check the if the final max is greater than its index)
            if (i<=max) {
                max = Math.max(max, i+nums[i]);
            } else {
                return false;
            }
        }
        return true;
    }
}